cmd/juju/application/deploy.go

```diff
@@ -1164,66 +1164,72 @@ func (c *DeployCommand) maybePredeployedLocalCharm() (deployFn, error) {
 	}, nil
 }
 
-func (c *DeployCommand) maybeReadLocalBundle() (deployFn, error) {
-	bundleFile := c.CharmOrBundle
-	var bundleDir string
-	isDir := false
-	resolveDir := false
-
+// readLocalBundle returns the bundle data and bundle dir (for
+// resolving includes) for the bundleFile passed in. If the bundle
+// file doesn't exist we return nil.
+func readLocalBundle(ctx *cmd.Context, bundleFile string) (*charm.BundleData, string, error) {
```

Comment:
We should return NotFoundError and have the callers deal with that if the bundle doesn't exist.
That's a lot more obvious and in line with expected semantics

Reply:
Yeah, I agree - the top-level functions in the deploy fallback pipeline can still return nil to indicate "try the next one", but that shouldn't happen further down the stack. Fixing.

Reply:
I used NotValid rather than NotFound - partly because that seems more correct and partly because it's less likely to be a real error that comes back.